client/transaction-pool/src/graph/base_pool.rs

```diff
@@ -395,33 +395,41 @@ impl<Hash: hash::Hash + Member + Serialize, Ex: std::fmt::Debug> BasePool<Hash,
 
 		while ready.is_exceeded(self.ready.len(), self.ready.bytes()) {
 			// find the worst transaction
-			let minimal = self.ready.fold(|minimal, current| {
+			let worst = self.ready.fold::<TransactionRef<Hash, Ex>, _>(|worst, current| {
 				let transaction = &current.transaction;
-				match minimal {
-					None => Some(transaction.clone()),
-					Some(ref tx) if tx.insertion_id > transaction.insertion_id =>
-						Some(transaction.clone()),
-					other => other,
-				}
+				worst
+					.map(|worst| {
+						match worst.transaction.priority.cmp(&transaction.transaction.priority) {
+							Ordering::Less => worst,
+							Ordering::Equal =>
+								if worst.insertion_id > transaction.insertion_id {
+									transaction.clone()
+								} else {
+									worst
+								},
+							Ordering::Greater => transaction.clone(),
+						}
+					})
+					.or_else(|| Some(transaction.clone()))
```

Comment:
```suggestion
if worst.is_none() || worst.as_ref() > Some(&current.transaction) {
  Some(current.transaction.clone())
} else {
  worst
}
```

It should be fine to re-use `Ord` implementation for `TransactionRef` which takes `validity` into account as well.

Reply:
I saw that, but I wasn't convinced `valid_till` should be taken into account here. I can update if you think that is more appropriate though.

Reply:
What's your concern with `valid_till`? I feel the rules should be exactly the same as for transaciton prioritisation (i.e. we remove the one that is the less likely to get in).

Reply:
The only concern I have is that if I understand it correctly, `valid_till` is based on longevity, which in turn is `u64::MAX` by default, which in turn means transactions with non-`u64::MAX` longevity will be evicted from transaction pool the first thing (priority is `0` by default). If that isn't a concern (I'm not entirely sure it is) then yes, regular comparison can be used there.

Reply:
So longevity is going to be `u64::MAX` for `Immortal` transactions indeed. For `Mortal` (recommended anyway) the `longevity` should be in the 10^2-10^3 orders of magnitude. However if you take a look at the `Ord` implementation for `TransactionRef` we favour transactions with *lower* longevity - i.e. we don't want them to be evicted from the pool without block inclusion, hence we prioritise them more (in case the priority is the same obviously).

Hence I think the behavior here should match the ordering with have for block inclusion, where we would rather evict an immortal transaction from the pool than a mortal one.

Reply:
The order comparison implementation applies is inverted for some reason, this patch is necessary to fix it:
```patch
--- a/client/transaction-pool/src/graph/ready.rs	(revision 12067cd14bc300c65a76babd2a218706f54f3219)
+++ b/client/transaction-pool/src/graph/ready.rs	(date 1638467249641)
@@ -56,8 +56,8 @@
 		self.transaction
 			.priority
 			.cmp(&other.transaction.priority)
-			.then_with(|| other.transaction.valid_till.cmp(&self.transaction.valid_till))
-			.then_with(|| other.insertion_id.cmp(&self.insertion_id))
+			.then_with(|| self.transaction.valid_till.cmp(&other.transaction.valid_till))
+			.then_with(|| self.insertion_id.cmp(&other.insertion_id))
 	}
 }
```

Which in turn breaks 3 other existing tests. Without that patch existing test for dropping older transactions starts failing. Either way, it will be a change in behavior, not extension to support transaction priority.

Reply:
We can't apply that patch, use-cases for ordering are different.

For inclusion we want to highest priority and oldest. For dropping we want to drop lowest priority, but also oldest.

If you want I can add logic to account for `valid_until` on top of the logic already present in this PR, but we definitely can't reuse the same ordering here.

Reply:
Good point. Fair enough, let's leave as-is currently and only take priority and insertion time into account.

Reply:
Sure, added comment in latest commit